File: packages/cli-node/src/integrations.ts
// Integration-strip: byte-parity TS port of tools/strip-integrations.mjs.

import { spawnSync } from 'node:child_process';
import type { Integration } from './args.js';

export const INTEGRATION_NAMES: readonly Integration[] = [
  'graphify',
  'obsidian',
  'caveman',
  'agentmemory',
  'openwiki',
  'ponytail',
  'opencodereview',
];
const VALID = new Set<string>(INTEGRATION_NAMES);

/** Strip integration-gated blocks; keep those whose NAME appears in `active`. */
export function stripIntegrations(source: string, active: readonly Integration[]): string {
  const set = new Set<string>(active);
  const blockRe = /<!--\s*integration:([a-z]+)\s*-->([\s\S]*?)<!--\s*\/integration:\1\s*-->\n?/g;
  let stripped = source.replace(blockRe, (_match, name: string, body: string) => {
    if (!VALID.has(name)) throw new Error(`unknown integration in marker: ${name}`);
    return set.has(name) ? body : '';
  });
  stripped = stripped.replace(/<!--\s*prettier-ignore-(start|end)\s*-->\n?/g, '');
  return stripped.replace(/\n{3,}/g, '\n\n').replace(/^\n+/, '').replace(/\n+$/, '\n');
}

// --- Registry: docs + optional local installer per integration ------------

export interface IntegrationInfo {
  title: string;
  docsUrl: string;
  /** argv of a local install command, or null when there is no clean installer. */
  install: string[] | null;
}

/** Metadata for every integration: friendly name, official docs, installer. */
export const INTEGRATION_REGISTRY: Record<Integration, IntegrationInfo> = {
  graphify: { title: 'Graphify', docsUrl: 'https://graphify.net/', install: null },
  obsidian: { title: 'Obsidian', docsUrl: 'https://obsidian.md/', install: null },
  caveman: {
    title: 'Caveman',
    docsUrl: 'https://github.com/JuliusBrussee/caveman',
    install: ['npx', '--yes', 'skills', 'add', 'JuliusBrussee/caveman'],
  },
  agentmemory: { title: 'agentmemory', docsUrl: 'https://github.com/rohitg00/agentmemory', install: null },
  openwiki: { title: 'openwiki', docsUrl: 'https://github.com/langchain-ai/openwiki', install: null },
  ponytail: { title: 'ponytail', docsUrl: 'https://github.com/DietrichGebert/ponytail', install: null },
  opencodereview: { title: 'Open Code Review', docsUrl: 'https://github.com/alibaba/open-code-review', install: null },
};

export interface InstallResult {
  name: Integration;
  status: 'installed' | 'failed' | 'manual' | 'skipped';
  command?: string;
}

export interface InstallOpts {
  isTty: boolean;
  noInstall: boolean;
  /** Injectable runner for tests; defaults to spawnSync inheriting stdio. */
  run?: (argv: string[]) => boolean;
}

function defaultRun(argv: string[]): boolean {
  const r = spawnSync(argv[0] as string, argv.slice(1), { stdio: 'inherit' });
  return r.status === 0;
}

/**
 * Run the known installer for each active integration that has one.
 * Skips all installs on a non-TTY run or when `noInstall` is set (prints the
 * command as `manual` instead). Never throws - a failed install is reported,
 * not fatal, so the scaffold still succeeds.
 */
export function installIntegrations(
  active: readonly Integration[],
  opts: InstallOpts,
): InstallResult[] {
  const run = opts.run ?? defaultRun;
  const out: InstallResult[] = [];
  for (const name of active) {
    const info = INTEGRATION_REGISTRY[name];
    if (!info.install) continue; // link-only integration
    const command = info.install.join(' ');
    if (opts.noInstall || !opts.isTty) {
      out.push({ name, status: 'manual', command });
      continue;
    }
    process.stdout.write(`\nInstalling ${info.title}: ${command}\n`);
    out.push({ name, status: run(info.install) ? 'installed' : 'failed', command });
  }
  return out;
}

File: packages/cli-python/src/spec_init/integrations.py
"""Strip integration-gated blocks; byte-parity Python port of strip-integrations.mjs."""

from __future__ import annotations

import re
import subprocess
import sys
from collections.abc import Callable
from dataclasses import dataclass
from typing import Iterable

from .args import Integration

INTEGRATION_NAMES: tuple[Integration, ...] = (
    "graphify",
    "obsidian",
    "caveman",
    "agentmemory",
    "openwiki",
    "ponytail",
    "opencodereview",
)
_VALID = frozenset(INTEGRATION_NAMES)
_BLOCK_RE = re.compile(
    r"<!--\s*integration:([a-z]+)\s*-->(.*?)<!--\s*/integration:\1\s*-->\n?",
    re.DOTALL,
)


def strip_integrations(source: str, active: Iterable[Integration]) -> str:
    """Strip integration blocks; keep those whose NAME appears in `active`."""
    active_set = set(active)

    def repl(match: re.Match[str]) -> str:
        name = match.group(1)
        body = match.group(2)
        if name not in _VALID:
            raise ValueError(f"unknown integration in marker: {name}")
        return body if name in active_set else ""

    stripped = _BLOCK_RE.sub(repl, source)
    stripped = re.sub(r"<!--\s*prettier-ignore-(start|end)\s*-->\n?", "", stripped)
    stripped = re.sub(r"\n{3,}", "\n\n", stripped)
    stripped = re.sub(r"^\n+", "", stripped)
    stripped = re.sub(r"\n+$", "\n", stripped)
    return stripped


# --- Registry: docs + optional local installer per integration ------------


@dataclass(frozen=True)
class IntegrationInfo:
    """Friendly name, official docs, and optional local install command."""

    title: str
    docs_url: str
    install: tuple[str, ...] | None


INTEGRATION_REGISTRY: dict[Integration, IntegrationInfo] = {
    "graphify": IntegrationInfo("Graphify", "https://graphify.net/", None),
    "obsidian": IntegrationInfo("Obsidian", "https://obsidian.md/", None),
    "caveman": IntegrationInfo(
        "Caveman",
        "https://github.com/JuliusBrussee/caveman",
        ("npx", "--yes", "skills", "add", "JuliusBrussee/caveman"),
    ),
    "agentmemory": IntegrationInfo("agentmemory", "https://github.com/rohitg00/agentmemory", None),
    "openwiki": IntegrationInfo("openwiki", "https://github.com/langchain-ai/openwiki", None),
    "ponytail": IntegrationInfo("ponytail", "https://github.com/DietrichGebert/ponytail", None),
    "opencodereview": IntegrationInfo(
        "Open Code Review", "https://github.com/alibaba/open-code-review", None
    ),
}


@dataclass(frozen=True)
class InstallResult:
    """Outcome of trying to install one integration."""

    name: Integration
    status: str  # 'installed' | 'failed' | 'manual' | 'skipped'
    command: str = ""


def _default_run(argv: tuple[str, ...]) -> bool:
    return subprocess.run(list(argv), check=False).returncode == 0


def install_integrations(
    active: Iterable[Integration],
    is_tty: bool,
    no_install: bool,
    run: Callable[[tuple[str, ...]], bool] | None = None,
) -> list[InstallResult]:
    """Run the known installer for each active integration that ships one.

    Skips all installs on a non-TTY run or when `no_install` is set (reports the
    command as `manual`). Never raises - a failed install is reported, not fatal.
    """
    runner = run or _default_run
    out: list[InstallResult] = []
    for name in active:
        info = INTEGRATION_REGISTRY[name]
        if info.install is None:
            continue  # link-only integration
        command = " ".join(info.install)
        if no_install or not is_tty:
            out.append(InstallResult(name=name, status="manual", command=command))
            continue
        sys.stdout.write(f"\nInstalling {info.title}: {command}\n")
        ok = runner(info.install)
        out.append(InstallResult(name=name, status="installed" if ok else "failed", command=command))
    return out
